homeassistant/components/media_player/hegel.py

```diff
@@ -0,0 +1,244 @@
+"""
+Support for Hegel IP Controlable Integrated Amplifiers.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.hegel/
+"""
+import logging
+import telnetlib
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    PLATFORM_SCHEMA, SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_OFF, SUPPORT_TURN_ON, SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET, MediaPlayerDevice)
+from homeassistant.const import (
+    CONF_HOST, CONF_NAME, CONF_PORT, STATE_OFF, STATE_ON,
+    STATE_UNKNOWN)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'Hegel Integrated Amplifier'
+DEFAULT_PORT = 50001   # telnet default port for Hegel devices
+
+SUPPORT_HEGEL = SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE | \
+                  SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Hegel platform."""
+    hegel = HegelDevice(
+        config.get(CONF_NAME), config.get(CONF_HOST), config.get(CONF_PORT))
+
+    if hegel.update():
+        add_entities([hegel])
+
+
+class HegelDevice(MediaPlayerDevice):
+    """Representation of a Hegel device."""
+
+    def __init__(self, name, host, port):
+        """Initialize the Hegel device."""
+        self._name = name
+        self._host = host
+        self._port = port
+        self._power_state = 1
+        self._volume = 0
+        self._muted = False
+        self._selected_source = 1
+        self._telnet_session = None
+        self._source_names = [
+            "Balanced",
+            "Analog1",
+            "Analog2",
+            "Coaxial",
+            "Optical1",
+            "Optical2",
+            "Optical3",
+            "USB",
+            "Network"
+        ]
+
+    def telnet_connect(self):
+        """Establish a telnet connection or return an already opened one."""
+        if self._telnet_session is None:
+            try:
+                self._telnet_session = telnetlib.Telnet(self._host, self._port)
+                _LOGGER.debug("Telnet connection established")
+            except (ConnectionRefusedError, OSError):
+                _LOGGER.warning("Connection refused by %s on port %s",
+                                self._host, self._port)
+                return False
+
+        return self._telnet_session
+
+    def telnet_disconnect(self):
+        """Close the telnet connection."""
+        telnet = self.telnet_connect()
+
+        if telnet:
+            telnet.close()
+            self._telnet_session = None
+        else:
+            return None
+
+    def telnet_send(self, command, parameter):
```

Comment:
All of this telnet logic needs to be moved to a separate library hosted on pypi. 

See [the checklist](https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services) for more details.

Reply:
I'll defer to @home-assistant/core for a final decision, but as far as I know, those platforms predate the requirement for all communication code to live in a library.

Reply:
Ok clear. Thanks! :)

Reply:
Things slip through. Al new platforms need to follow the rules.